packages/block-editor/src/hooks/layout.js

```diff
@@ -354,76 +418,63 @@ export const withInspectorControls = createHigherOrderComponent(
  */
 export const withLayoutStyles = createHigherOrderComponent(
 	( BlockListBlock ) => ( props ) => {
-		const { name, attributes } = props;
-		const blockSupportsLayout = hasLayoutBlockSupport( name );
-		const disableLayoutStyles = useSelect( ( select ) => {
-			const { getSettings } = select( blockEditorStore );
-			return !! getSettings().disableLayoutStyles;
-		} );
-		const shouldRenderLayoutStyles =
-			blockSupportsLayout && ! disableLayoutStyles;
-		const id = useInstanceId( BlockListBlock );
-		const { layout } = attributes;
-		const { default: defaultBlockLayout } =
-			getBlockSupport( name, layoutBlockSupportKey ) || {};
-		const usedLayout =
-			layout?.inherit || layout?.contentSize || layout?.wideSize
-				? { ...layout, type: 'constrained' }
-				: layout || defaultBlockLayout || {};
-		const layoutClasses = blockSupportsLayout
-			? useLayoutClasses( attributes, name )
-			: null;
-		// Higher specificity to override defaults from theme.json.
-		const selector = `.wp-container-${ id }.wp-container-${ id }`;
-		const [ blockGapSupport ] = useSettings( 'spacing.blockGap' );
-		const hasBlockGapSupport = blockGapSupport !== null;
-
-		// Get CSS string for the current layout type.
-		// The CSS and `style` element is only output if it is not empty.
-		let css;
-		if ( shouldRenderLayoutStyles ) {
-			const fullLayoutType = getLayoutType(
-				usedLayout?.type || 'default'
-			);
-			css = fullLayoutType?.getLayoutStyle?.( {
-				blockName: name,
-				selector,
-				layout: usedLayout,
-				style: attributes?.style,
-				hasBlockGapSupport,
-			} );
+		if ( ! hasLayoutBlockSupport( props.name ) ) {
+			return <BlockListBlock { ...props } />;
 		}
 
-		// Attach a `wp-container-` id-based class name as well as a layout class name such as `is-layout-flex`.
-		const layoutClassNames = classnames(
-			{
-				[ `wp-container-${ id }` ]: shouldRenderLayoutStyles && !! css, // Only attach a container class if there is generated CSS to be attached.
-			},
-			layoutClasses
-		);
-
-		const { setStyleOverride, deleteStyleOverride } = unlock(
-			useDispatch( blockEditorStore )
-		);
-
-		useEffect( () => {
-			if ( ! css ) return;
-			setStyleOverride( selector, { css } );
-			return () => {
-				deleteStyleOverride( selector );
-			};
-		}, [ selector, css, setStyleOverride, deleteStyleOverride ] );
-
 		return (
-			<BlockListBlock
-				{ ...props }
-				__unstableLayoutClassNames={ layoutClassNames }
+			<BlockListWithLayoutStyles
+				block={ BlockListBlock }
+				props={ props }
 			/>
 		);
 	},
 	'withLayoutStyles'
 );
 
+function BlockListWithChildLayoutStyles( { block: BlockListBlock, props } ) {
+	const { style: { layout = {} } = {} } = props.attributes;
+	const { selfStretch, flexSize } = layout;
+	const disableLayoutStyles = useSelect( ( select ) => {
+		const { getSettings } = select( blockEditorStore );
+		return !! getSettings().disableLayoutStyles;
+	} );
```

Comment:
The `disableLayoutStyles` could be efficiently moved into the parent HOC component, leading to rendering the `BlockWithLayoutStyles` component less often. Just use this pattern:
```js
const hasChildLayout = ...;
const shouldRenderLayoutStyles = useSelect( ( select ) => {
  if ( ! hasChildLayout ) {
    return false; // return early before calling `select` at all
  }
  return ! select( blockEditorStore ).getSettings().disableLayoutStyles;
}, [ hasChildLayout ] );
```
The trick is that while `hasChildLayout` remains `false`, the `select` function is not called and no store subscriptions are created.

Reply:
Great suggestion! I forgot about this pattern, even though I read your previous PRs 😓